### PR TITLE
HDDS-3955. Unable to list intermediate paths on keys created using S3G.

### PR DESCRIPTION
**Pull request by bharatviswa504**
## What changes were proposed in this pull request?

Keys created via the S3 Gateway currently use the createKey OM API to create the ozone key. Hence, when using a hdfs client to list intermediate directories in the key, OM returns key not found error. This was encountered while using fluentd to write Hive logs to Ozone via the s3 gateway.

Added a new config in OM to control the behavior of key create to create intermediate directories or not. In this way, keys created via ozone S3G can be accessed through fs API.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3955

## How was this patch tested?

Added UT and IT.

### CODE REVIEW

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java [hunk 1]
```diff
@@ -239,4 +239,11 @@ private OMConfigKeys() {
           "ozone.om.keyname.character.check.enabled";
   public static final boolean OZONE_OM_KEYNAME_CHARACTER_CHECK_ENABLED_DEFAULT =
           false;
+
+  // This config needs to be enabled, when S3G created objects will be used
+  // FileSystem.
```
Comment (arp7) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java:244]: used via FileSystem API.

**Comment by arp7** [comment]
Can you update the PR template a bit to describe the high-level approach (was fix made in OM or S3G)? 🙂

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java [hunk 2]
```diff
@@ -221,8 +233,45 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
...
+        OMFileRequest.OMPathInfo pathInfo =
+            OMFileRequest.verifyFilesInPath(omMetadataManager, volumeName,
+                bucketName, keyName, Paths.get(keyName));
+        OMFileRequest.OMDirectoryResult omDirectoryResult =
+            pathInfo.getDirectoryResult();
+        inheritAcls = pathInfo.getAcls();
+
+        // Check if a file or directory exists with same key name.
+        if (omDirectoryResult == DIRECTORY_EXISTS) {
+          throw new OMException("Can not write to directory: " + keyName,
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:253]: Let's make the exception message slightly more descriptive to explain what is really going on. We can mention that createIntermediateDirs behavior is enabled and hence `/` has special interpretation.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java [hunk 3]
```diff
@@ -221,8 +233,45 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
...
+          if (omDirectoryResult == FILE_EXISTS_IN_GIVENPATH) {
+            throw new OMException("Can not create file: " + keyName +
+                " as there is already file in the given path", NOT_A_FILE);
+          }
+
+        missingParentInfos = OMDirectoryCreateRequest
+            .getAllParentInfo(ozoneManager, keyArgs,
+                pathInfo.getMissingParents(), inheritAcls, trxnLogIndex);
+
+        // Add cache entries for the prefix directories.
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:265]: @bharatviswa504 , one q. We are adding cache entries right away instead of at the end on success. Could this be a potential issue on the failure path?

**Comment by arp7** [comment]
The approach looks really good. It is surprisingly concise and elegant because I was expecting it to be a lot more code! Added a few review comments.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java [hunk 4]
```diff
@@ -328,6 +342,73 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
...
+        100L, ozoneManagerDoubleBufferHelper);
+
+    Assert.assertEquals(omClientResponse.getOMResponse().getStatus(), OK);
+
+    Path keyPath = Paths.get(keyName);
+
+    // Check intermediate paths are created
+    keyPath = keyPath.getParent();
+    while(keyPath != null) {
+      Assert.assertNotNull(omMetadataManager.getKeyTable().get(
```
Comment (mukul1987) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:377]: Can we change this to fs.getFileStatus(keypath) so that returns status.isDir() ?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java (hunk 3 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:265]: Yes, there will be some intermediate directories left in DB, key commit failed. There will be intermediate directories created.
But I don't see that causing an issue. But cleaner thing is cleaning up entries. 
But this is a common issue for FileCreate and KeyCreate.

And in the case lets say key create itself failed in DB flush, we terminate DB.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java (hunk 2 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java]: Done

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java (hunk 1 above)
Comment (bharatviswa504) [reply hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java]: done

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java (hunk 4 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:376]: Added test

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 5]
```diff
@@ -265,4 +272,24 @@ public AuditMessage buildAuditMessage(AuditAction op,
...
+    if (enableFileSystemPaths) {
+      String normalizedKeyName;
+      if (keyName.startsWith(OM_KEY_PREFIX)) {
+        normalizedKeyName = Paths.get(keyName).toUri().normalize().getPath();
+      } else {
+        normalizedKeyName = Paths.get(OM_KEY_PREFIX, keyName).toUri()
+            .normalize().getPath();
+      }
+      if (!keyName.equals(normalizedKeyName)) {
+        LOG.debug("Normalized key {} to {} ", keyName, normalizedKeyName);
```
Comment (avijayanhwx) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:288]: Can we print normalizedKeyName.substring(1) given we return that?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 6]
```diff
@@ -265,4 +272,24 @@ public AuditMessage buildAuditMessage(AuditAction op,
     auditMap.put(OzoneConsts.VOLUME, volume);
     return auditMap;
   }
+
+  @SuppressFBWarnings("DMI_HARDCODED_ABSOLUTE_FILENAME")
+  public static String getNormalizedKey(boolean enableFileSystemPaths,
```
Comment (avijayanhwx) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:277]: What about moving the 'enableFileSystemPaths' flag outside, and call the method conditionally? In that way, the getNormalizedKey will stay a simple util function without knowing about application logic.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+        getNormalizedKey(true, "/a/b/c/d"));
+    Assert.assertEquals("a/b/c/d",
+        getNormalizedKey(true, "////a/b/c/d"));
+    Assert.assertEquals("a/b/c/d",
+        getNormalizedKey(true, "////a/b/////c/d"));
+    Assert.assertEquals("a/b/c/...../d",
+        getNormalizedKey(true, "////a/b/////c/...../d"));
+    Assert.assertEquals("a/b/d",
+        getNormalizedKey(true, "/a/b/c/../d"));
+    Assert.assertEquals("a/d",
```
Comment (avijayanhwx) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java:47]: Nit. Can we add a test for key with length = 1? Since we return .substring(1) in the method, trying to be cautious here.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java (hunk 6 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:277]: That requires in all over the code, we need to check this flag and call this. I feel this is better way.
If needed I can move the code of actual Normalize, to another utility if you expect this to be used else where.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java (hunk 7 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java:47]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java (hunk 6 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:277]: Moved common logic to a new method.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java (hunk 5 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:288]: Done

File: hadoop-hdds/common/src/main/resources/ozone-default.xml [hunk 8]
```diff
@@ -2478,4 +2478,18 @@
       rules in Amazon S3's object key naming guide.
     </description>
   </property>
+
+  <property>
+    <name>ozone.om.enable.filesystem.paths</name>
+    <tag>OZONE, OM</tag>
+    <value>false</value>
+    <description>If true, OM will normalize the key paths and also create
```
Comment (arp7) [inline hadoop-hdds/common/src/main/resources/ozone-default.xml:2486]: I would rephrase first sentence a bit to state that key names will be interpreted as file system paths. `/` will be treated as a special character and paths will be normalized and must follow Unix filesystem path naming conventions.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 9]
```diff
@@ -265,4 +272,30 @@ public AuditMessage buildAuditMessage(AuditAction op,
...
+  public static String getNormalizedKey(String keyName) {
+    String normalizedKeyName;
+    if (keyName.startsWith(OM_KEY_PREFIX)) {
+      normalizedKeyName = Paths.get(keyName).toUri().normalize().getPath();
+    } else {
+      normalizedKeyName = Paths.get(OM_KEY_PREFIX, keyName).toUri()
+          .normalize().getPath();
+    }
+    if (!keyName.equals(normalizedKeyName)) {
+      LOG.debug("Normalized key {} to {} ", keyName,
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:296]: Thanks for adding this log message. It will be super useful!

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java [hunk 10]
```diff
@@ -79,15 +79,11 @@ public static OMPathInfo verifyFilesInPath(
...
-      if (omMetadataManager.getKeyTable().isExist(dbKeyName)) {
-        // Found a file in the given path.
-        // Check if this is actual file or a file in the given path
-        if (dbKeyName.equals(fileNameFromDetails)) {
-          result = OMDirectoryResult.FILE_EXISTS;
-        } else {
-          result = OMDirectoryResult.FILE_EXISTS_IN_GIVENPATH;
-        }
-      } else if (omMetadataManager.getKeyTable().isExist(dbDirKeyName)) {
+      // Check first for dir. This is to handle leading "/" in path, which
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:82]: Leading `/` in path should be dropped at this point, right?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java [hunk 11]
```diff
@@ -79,15 +79,11 @@ public static OMPathInfo verifyFilesInPath(
       String dbDirKeyName = omMetadataManager.getOzoneDirKey(volumeName,
           bucketName, pathName);
 
-      if (omMetadataManager.getKeyTable().isExist(dbKeyName)) {
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:82]: I am a little bit worried about changes to OMFileRequest. There is a risk of changing the behavior and invalidating the app-compat testing we have done so far.

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:82]: This change can be reverted as now we don't allow leading / in path names when OMConfig enable.filesystems is set to true. Will updated in my next PR update.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java (hunk 10 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:82]: Yes. This change was required, when we are allowing leading / in keyName. I will revert the change

**Comment by bharatviswa504** [comment]
Thank You @arp7 and @avijayanhwx for the review.
I have addressed review comments.

**Comment by bharatviswa504** [comment]
Rebased with latest apache main branch.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java [hunk 12]
```diff
@@ -265,4 +273,72 @@ public AuditMessage buildAuditMessage(AuditAction op,
...
+      for (int i = 0; i < components.length; i++) {
+        String element = components[i];
+        if (element.equals(".") ||
+            (element.contains(":")) ||
+            (element.contains("/") || element.equals(".."))) {
+          isValid = false;
+          break;
+        }
+
+        // The string may end with a /, but not have
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:330]: Does this mean we are allowing trailing `/`?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequest.java [hunk 13]
```diff
@@ -141,7 +141,6 @@ public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
               ozoneManager.getPreallocateBlocksMax(),
               ozoneManager.isGrpcBlockTokenEnabled(),
               ozoneManager.getOMNodeId());
-
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequest.java:144]: Spurious change?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java (hunk 12 above)
Comment (arp7) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:330]: I realized this is checked later in `OmKeyCreateRequest`.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java [hunk 14]
```diff
@@ -91,6 +99,19 @@ public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
...
+
+    String keyPath = keyArgs.getKeyName();
+    if (ozoneManager.getEnableFileSystemPaths()) {
+      // If enabled, disallow keys with trailing /. As in fs semantics
+      // directories end with trailing /.
+      keyPath = validateAndNormalizeKey(
+          ozoneManager.getEnableFileSystemPaths(), keyPath);
+      if (keyPath.endsWith("/")) {
+        throw new OMException("Invalid KeyPath: " + keyPath,
+            OMException.ResultCodes.INVALID_KEY_NAME);
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:111]: Let's also mention why it is invalid, else user may be scratching their head. 🙂

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java [hunk 15]
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
...
+
+    // converted to a/b/c/.../file3
+    keyName = "///a/b///c//.../file3";
+    createAndCheck(keyName);
+
+    // converted to r1/r2
+    keyName = "././r1/r2/";
+    createAndCheck(keyName);
+
+    // converted to ..d1/d2/d3
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:390]: Why is this converted to `../d1/d2/d3`? Should it be failed instead due to trailing `/`?

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java [hunk 16]
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
...
+    checkNotAValidPath(keyName);
+
+    // Key name ends with /
+    keyName = "/a/./";
+    checkNotAValidPath(keyName);
+
+    keyName = "/////";
+    checkNotAValidPath(keyName);
+
+    keyName = "../../b/c";
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:410]: Also let's add a check for `../a/b`.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java [hunk 17]
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
...
+            101L, ozoneManagerDoubleBufferHelper);
+
+    Assert.assertEquals(OK, omClientResponse.getOMResponse().getStatus());
+
+    checkCreatedPaths(omKeyCreateRequest, omRequest, keyName);
+  }
+
+  private void checkCreatedPaths(OMKeyCreateRequest omKeyCreateRequest,
+      OMRequest omRequest, String keyName) throws Exception {
+    keyName = omKeyCreateRequest.validateAndNormalizeKey(true, keyName);
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:482]: Let's add a separate test case for `validateAndNormalizeKey` itself. All it does is call `validateAndNormalizeKey` with bunch of hard-coded input and asserts on the expected output.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java [hunk 18]
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
...
+    // Check intermediate directories created or not.
+    Path keyPath = Paths.get(keyName);
+    checkIntermediatePaths(keyPath);
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:485]: Nice!

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java (hunk 17 above)
Comment (arp7) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:482]: Ignore this comment, I see it above now. Was reviewing files out of order.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+    Assert.assertEquals("a/b/c/d",
+        validateAndNormalizeKey(true, "a/b/c/d"));
+    Assert.assertEquals("a/b/c/d",
+        validateAndNormalizeKey(true, "/a/b/c/d"));
+    Assert.assertEquals("a/b/c/d",
+        validateAndNormalizeKey(true, "////a/b/c/d"));
+    Assert.assertEquals("a/b/c/d",
+        validateAndNormalizeKey(true, "////a/b/////c/d"));
+    Assert.assertEquals("a/b/c/...../d",
+        validateAndNormalizeKey(true, "////a/b/////c/...../d"));
```
Comment (arp7) [inline hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java:51]: Bharat, more than two dots should normalize to itself. In Unix I can create a dir named `...` or `.....`.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java (hunk 12 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:330]: Yes, we don't allow for a keyName with trailing /. This method can return after normalization in some cases. So, we checked in OmKeyCreateRequest

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java (hunk 19 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java:51]: yes, you are right. It is the samething right, we are checking a/b/c/...../d, just it has removed additional and leading /

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java (hunk 15 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:390]: ..d1/d2/d3/ -> ../d1/d2/d3 it removed trailing /. Normalize will take care of removing additional /

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:390]: we will fail, even after normalize if we have a case like /a/. -> a/

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java (hunk 16 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:410]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java (hunk 14 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:112]: Done

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequest.java (hunk 13 above)
Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequest.java]: Removed it

**Comment by bharatviswa504** [comment]
> Few minor comments, and couple of questions on the expected normalizations.
> 
> Overall the change looks good. Thanks for adding comprehensive test cases.

Thank You @arp7 for the review. addressed review comments.

File: hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java (hunk 19 above)
Comment (arp7) [reply hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java:51]: Right, I got confused.
